dali/python/nvidia/dali/pool.py

```diff
@@ -0,0 +1,269 @@
+    def _start_processes(self):
+        try:
+            for process in self.processes:
+                process.start()
+            from_tracker_r, from_tracker_w = multiprocessing.Pipe(duplex=False)
+            to_tracker_r, to_tracker_w = multiprocessing.Pipe(duplex=False)
+            self.from_tracker = from_tracker_r
+            self.to_tracker = to_tracker_w
+            self.tracker_thread = threading.Thread(
+                target=join_processes, args=(
+                    self.processes, to_tracker_r, from_tracker_w, self.task_pipes, self.task_pipes_lock
+                )
+            )
+            self.tracker_thread.start()
+        except:
+            for proc in self.processes:
+                if proc.is_alive():
+                    proc.terminate()
+            for proc in self.processes:
+                proc.join()
+            raise
+
+
+class WorkersPool(object):
+
+    def __init__(self, callbacks_no, pool):
+        self.contexts = [CallbackContext() for _ in range(callbacks_no)]
+        self.pool = pool
+        self.rec_pipes = self.pool.res_pipes + [self.pool.from_tracker]
+
+    @classmethod
+    def from_groups(cls, groups, workers_no, init_method, keep_alive_queue_size, initial_chunk_size=1024 * 1024):
+        callbacks = [group.callback for group in groups]
+        queue_depths = [keep_alive_queue_size + group.prefetch_queue_depth for group in groups]
+        pool = ProcPool(callbacks, workers_no, init_method, queue_depths, initial_chunk_size)
+        return cls(len(callbacks), pool)
+
+    def schedule_batch(self, context_i, batch_i, pipeline, tasks):
+        tasks = list(enumerate(tasks))
+        if not tasks:
+            raise RuntimeError("Cannot schedule empty list of tasks")
+        context = self.contexts[context_i]
+        if batch_i in context.scheduled:
+            raise RuntimeError("Given batch has already been scheduled")
+        self.distribute((context_i, batch_i), tasks)
+        context.partially_received[batch_i] = {}
+        context.scheduled[batch_i] = tasks
+
+    def distribute(self, header, tasks):
+        workers_no = self.pool.workers_no
+        tasks_no = len(tasks)
+        chunk_size = tasks_no // workers_no
+        queued_no = chunk_size + (tasks_no % workers_no)
+        with self.pool.task_pipes_lock:
+            task_pipes = self.pool.task_pipes
+            task_pipes[0].send((header, tasks[:queued_no]))
+            if chunk_size:
+                for worker_id in range(1, workers_no):
+                    task_pipes[worker_id].send((header, tasks[queued_no:queued_no + chunk_size]))
+                    queued_no += chunk_size
+
+    def receive_batch(self, context_i):
+        context = self.contexts[context_i]
+        batch_i, tasks = context.scheduled.popitem(last=False)
+        awaiting_batch = context.partially_received[batch_i]
+        while len(awaiting_batch) < len(tasks) and batch_i not in context.iter_ended:
+            self._receive_chunk()
```

Comment:
Random question: can we somehow get starved/blocked by a consuming batches after `batch_i` that come from other workers? I guess in the end, the "lazy" worker will send us the last part of the batch, and when receiving next batch we will already have most of it in that case, right?

Reply:
Yes, that's exactly the idea. Even if we receive some other parts along the way they will simply wait for being received by the right externalsource group in the right order. Worker will finally send it's part or will die, either way we should not wait forever.